### PR TITLE
Add functionality to switch search algorithms

### LINKED ISSUE #1342: Add search algorithm switching UI from Orangelight
See pulibrary/orangelight#3744

This will enable us to try new algorithms, like ones which will solve #1341.

## Implementation Details

We don't have FlipFlop, so we might have to use some sort of config to enable/disable this?

- [ ] update solr configs as in https://github.com/pulibrary/pul_solr/pull/375 and https://github.com/pulibrary/pul_solr/pull/377
- [ ] add backend https://github.com/pulibrary/orangelight/pull/3819
- [ ] add UI https://github.com/pulibrary/orangelight/pull/3827

As an example, boost results with the online facet.

### PR DESCRIPTION
**Pull request by eliotjordan**
Closes #1342 

Screenshot 2023-11-15 at 11 35 20 AM

### CODE REVIEW

File: app/controllers/catalog_controller.rb [hunk 1]
```diff
@@ -118,6 +119,7 @@ def index
     # config.show.thumbnail_field = 'thumbnail_path_ss'
 
     config.add_results_collection_tool(:sort_widget)
+    config.add_results_collection_tool(:algorithm_select)
```
Comment (tpendragon) [inline app/controllers/catalog_controller.rb:122]: Can we add this to the module somehow?

File: app/controllers/concerns/multiple_algorithms.rb [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+    return unless search_algorithm_param && allowed_search_algorithms.include?(search_algorithm_param)
+
+    "#{search_algorithm_param}_search_builder".camelize.constantize
+  end
+
+  # When adding a new Ranking Algorithm the name will need to be added to this list before it can be utilized
+  # For example if you added a cats ranking algorithm, with a CatsSearchBuilder you would add "cats" to this list ["engineering","cats"]
+  # This is to make sure the user can not just execute any SearchBuilder in the system
+  def allowed_search_algorithms
+    ["online_content"]
```
Comment (tpendragon) [inline app/controllers/concerns/multiple_algorithms.rb:19]: Maybe this should be overridden in the catalog controller? I'm trying to think about how to make maintenance of this be as copy/paste as possible.

File: app/helpers/pulfalight_helper.rb [hunk 3]
```diff
@@ -132,6 +132,15 @@ def regular_compact_breadcrumbs(document)
     )
   end
 
+  # Provides the search algorithm value for the UI component
+  def search_algorithm_value
```
Comment (tpendragon) [inline app/helpers/pulfalight_helper.rb:136]: Can we add default to allowed_search_algorithms and check that somehow?

**Review by tpendragon — CHANGES_REQUESTED** [review]
Just a couple of things I think we should think about as we imagine this feature existing in multiple apps

File: app/controllers/catalog_controller.rb (hunk 1 above)
Comment (eliotjordan) [reply app/controllers/catalog_controller.rb:122]: This is possible:
```

  def blacklight_config
    config = super
    config.add_results_collection_tool(:algorithm_select)
    config
  end
```

but it's not going to work out well. The location of the dropdown will vary depending on the app as each app will have different widgets enabled.

File: app/helpers/pulfalight_helper.rb (hunk 3 above)
Comment (eliotjordan) [reply app/helpers/pulfalight_helper.rb:136]: Done.

File: app/controllers/concerns/multiple_algorithms.rb (hunk 2 above)
Comment (eliotjordan) [reply app/controllers/concerns/multiple_algorithms.rb]: Done.

File: app/helpers/pulfalight_helper.rb (hunk 3 above)
Comment (eliotjordan) [reply app/helpers/pulfalight_helper.rb:136]: Done

**Pull request merged**